examples/algorithm_progression/xarray_sweep.py

```diff
@@ -0,0 +1,67 @@
+from pathlib import Path
+
+import numpy as np
+import xarray as xr
+from algorithm_sweep import AlgorithmSweep
+
+import large_image
+
+
+class AlgorithmSweepXArray(AlgorithmSweep):
+    def initialize_storage(self):
+        self.output_files = []
+
+    def apply_algorithm(self, param_combo):
+        iteration_filename = (
+            Path(self.output_dir) / f'result_{"_".join(str(v) for v in param_combo)}.nc'
+        )
+        if not iteration_filename.exists():
+            current_coords = {
+                k: [param_combo[i]] for i, k in enumerate(self.input_params.keys())
+            }
+            source = large_image.open(self.input_filename)
+            source_metadata = source.getMetadata()
+            output_data = np.empty(
+                (
+                    source_metadata["sizeY"],
+                    source_metadata["sizeX"],
+                    4,
+                )
+            )
+            for tile in source.tileIterator(
+                format=large_image.tilesource.TILE_FORMAT_NUMPY,
+                tile_size=dict(width=2048, height=2048),
+            ):
+                output_data[
+                    tile["y"]: tile["y"] + tile["height"],
+                    tile["x"]: tile["x"] + tile["width"],
+                ] = self.algorithm(tile["tile"], *param_combo)
+
+            xr.DataArray(
+                output_data,
+                dims=["y", "x", "c"],
+                attrs=current_coords,
+            ).to_netcdf(iteration_filename)
+        return iteration_filename
+
+    def handle_result(self, result):
+        self.output_files.append(result)
+
+    def complete_storage(self):
+        results = [xr.open_dataarray(filename) for filename in self.output_files]
+        if len(results) > 0:
+            dims = dict(
+                **{k: len(v) for k, v in self.input_params.items()},
+                **results[0].sizes,
+            )
+            composite_xr = xr.DataArray(
```

Comment:
It looks like this is putting everything in memory, so it fails if the results are larger than memory+swap.  Is there a way to ask xarray to not have it all in memory?  My test was to ask for a data set that was larger than my memory size and the process had to be OOM killed.

Reply:
We would have to store the array as zarr to do so:
https://discourse.pangeo.io/t/processing-large-too-large-for-memory-xarray-datasets-and-writing-to-netcdf/1724

Reply:
One of the fundamental problems we're facing here is the idea of editing the same data store from multiple processes. For the multisource sweep, this worked because each iteration writes its own file and the composite file (`results.yml`) is pretty small. When we try this for the xarray or hdf5 sweep, we want to end up with one large file. I haven't found a solution yet that allows chunked access of the file from multiple processes. What do you think is a good pattern to follow?